cmd/neutree-core/app/builder.go

```diff
@@ -0,0 +1,129 @@
+package app
+
+import (
+	"fmt"
+
+	"github.com/neutree-ai/neutree/cmd/neutree-core/app/config"
+	"github.com/neutree-ai/neutree/controllers"
+)
+
+// Builder is the application builder
+type Builder struct {
+	controllerInits map[string]ControllerFactory
+	config          *config.CoreConfig
+
+	beforeHooks map[string][]controllers.HookFunc
+	afterHooks  map[string][]controllers.HookFunc
+}
+
+// NewBuilder creates a new CLI builder
+func NewBuilder() *Builder {
+	b := &Builder{
+		controllerInits: make(map[string]ControllerFactory),
+		beforeHooks:     make(map[string][]controllers.HookFunc),
+		afterHooks:      make(map[string][]controllers.HookFunc),
+	}
+
+	defaultControllers := map[string]ControllerFactory{
+		"cluster":         NewClusterControllerFactory(),
+		"engine":          NewEngineControllerFactory(),
+		"endpoint":        NewEndpointControllerFactory(),
+		"role":            NewRoleControllerFactory(),
+		"role-assignment": NewRoleAssignmentControllerFactory(),
+		"workspace":       NewWorkspaceControllerFactory(),
+		"api-key":         NewApiKeyControllerFactory(),
+		"image-registry":  NewImageRegistryControllerFactory(),
+		"model-catalog":   NewModelCatalogControllerFactory(),
+		"model-registry":  NewModelRegistryControllerFactory(),
+	}
+
+	for name, factory := range defaultControllers {
+		b.controllerInits[name] = factory
+	}
+
+	return b
+}
+
+func (b *Builder) WithConfig(c *config.CoreConfig) *Builder {
+	b.config = c
+	return b
+}
+
+// WithController registers a controller factory
+func (b *Builder) WithController(name string, factory ControllerFactory) *Builder {
+	b.controllerInits[name] = factory
+	return b
+}
+
+func (b *Builder) WithGlobalBeforeReconcileHook(hook controllers.HookFunc) *Builder {
```

Comment:
This seems like a potential pitfall, because if you call the with global method before registering the controller, the global hook won't take effect.

Do we currently need a global hook?

Reply:
The issue with the global hook implementation has been fixed.
From my perspective, I think global hooks are necessary. We can use this mechanism to extend some common capabilities without registering the same hook for different controllers.

Reply:
ok